### ISSUE TITLE
Error waiting for internet gateway to detach / Cluster has node groups attached

### ISSUE THREAD

**Issue by richstokes** [terraform-providers/terraform-provider-aws#10934]
### Terraform Version
Terraform v0.12.16
+ provider.aws v2.38.0
+ provider.http v1.1.1

### Affected Resource(s)
* aws_internet_gateway

### Terraform Configuration Files
```hcl
resource "aws_internet_gateway" "demo" {
  vpc_id = aws_vpc.demo.id

  tags = {
    Name = "terraform-eks-demo"
  }
}
```

### Expected Behavior

AWS internet gateway should have been deleted

### Actual Behavior

```
aws_internet_gateway.demo: Still destroying... [id=igw-redacted, 14m50s elapsed]

Error: error deleting EKS Cluster (terraform-eks-demo): ResourceInUseException: Cluster has node groups attached
	status code: 409, request id: 1d511535-xxxx-4551-9cf3-759ef47cf8ce

Error: Error waiting for internet gateway (igw-redacted) to detach: timeout while waiting for state to become 'detached' (last state: 'detaching', timeout: 15m0s)
```

### Steps to Reproduce

1. `terraform destroy`

### Important Factoids

The node group its complaining about was also created via terraform. Looks like some kind of race condition as terraform should destroy the node group first.

I also see this error on the AWS side:  
```
"errorMessage": "Network vpc-redacted has some mapped public address(es). Please unmap those public address(es) before detaching the gateway."
```

Shouldn't terraform take care of any unmappings? Since terraform created them all..?

Here's the nodegroup resource if needed:

```hcl
resource "aws_eks_node_group" "nodegroup0" {
  cluster_name    = var.cluster-name
  node_group_name = "nodegroup0"
  node_role_arn   = aws_iam_role.demo-node.arn
  subnet_ids      = aws_subnet.demo[*].id

  disk_size       = 50
  instance_types  = ["r5.large"]

  version         = "1.14" # Kubernetes version

  scaling_config {
    desired_size = 2
    max_size     = 100
    min_size     = 2
  }
}
```

The rest of the code is following the example here: https://github.com/terraform-providers/terraform-provider-aws/tree/master/examples/eks-getting-started

**Comment by richstokes** [comment]
Running the destroy command again its also stuck removing the subnets:

```
Error: Error deleting subnet: timeout while waiting for state to become 'destroyed' (last state: 'pending', timeout: 20m0s)


Error: Error deleting subnet: timeout while waiting for state to become 'destroyed' (last state: 'pending', timeout: 20m0s)


Error: Error waiting for internet gateway (igw-0ec3fdd0877f740f6) to detach: timeout while waiting for state to become 'detached' (last state: 'detaching', timeout: 15m0s)
```

**Comment by richstokes** [comment]
Looks like the issue may be Kubernetes creating load balancers and elastic IPs in the same VPC

**Comment by bflad** [comment]
Hi folks 👋 Sorry for the trouble here.

If your Terraform configuration manages the IAM Role and its permissions for the EKS Node Group in the same configuration as the EKS Node Group, the `aws_eks_node_group` resource needs to include explicit dependencies on the IAM Role (inline or attached) policy resources in Terraform, e.g.

```hcl
resource "aws_eks_node_group" "example" {
  # ... other configuration ...

  # Ensure that IAM Role permissions are created before and deleted after EKS Node Group handling.
  # Otherwise, EKS will not be able to properly delete EC2 Instances and Elastic Network Interfaces.
  depends_on = [
    aws_iam_role_policy_attachment.example-AmazonEKSWorkerNodePolicy,
    aws_iam_role_policy_attachment.example-AmazonEKS_CNI_Policy,
    aws_iam_role_policy_attachment.example-AmazonEC2ContainerRegistryReadOnly,
  ]
}
```

This ensures that EKS still has the proper permissions to remove EC2 Instances and Elastic Network Interfaces (the likely cause of `DependencyViolation` errors for the `aws_subnet` resource during destroy in this situation). I have submitted #11010 to update the `aws_eks_node_group` resource documentation to note these explicit dependencies.

In the future, we can help reduce the need for explicit Terraform dependencies such as these via supporting the management of attached IAM Role policies directly in the `aws_iam_role` resource (e.g. #5904). 👍

**Comment by bflad** [comment]
The documentation update mentioned above has been merged and will release with version 2.40.0 of the Terraform AWS Provider later this week. 👍 

If you are still having issues after setting explicit dependencies between the EKS Node Group and its IAM Role policies, please submit a new [bug report](https://github.com/terraform-providers/terraform-provider-aws/issues/new?template=Bug_Report.md) and we can take a further look. Thanks!
